pkg/reconciler/apis/replicationclusterrolebinding/replicationclusterrolebinding_controller.go

```diff
@@ -0,0 +1,254 @@
+// 1. either a maximum-permission-policy subject is bound
+// 2. or a ClusterRole
+type controller struct {
+	queue workqueue.RateLimitingInterface
+
+	kubeClusterClient kcpkubernetesclientset.ClusterInterface
+
+	clusterRoleBindingLister  kcprbaclisters.ClusterRoleBindingClusterLister
+	clusterRoleBindingIndexer cache.Indexer
+
+	clusterRoleLister  kcprbaclisters.ClusterRoleClusterLister
+	clusterRoleIndexer cache.Indexer
+}
+
+// enqueueClusterRoleBinding enqueues an ClusterRoleBinding.
+func (c *controller) enqueueClusterRoleBinding(obj interface{}, values ...interface{}) {
+	key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	logger := logging.WithQueueKey(logging.WithReconciler(klog.Background(), ControllerName), key)
+	logger.V(4).WithValues(values...).Info("queueing ClusterRoleBinding")
+	c.queue.Add(key)
+}
+
+func (c *controller) enqueueClusterRole(obj interface{}) {
+	key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	_, _, name, err := kcpcache.SplitMetaClusterNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	objs, err := c.clusterRoleBindingIndexer.ByIndex(replicationclusterrole.ClusterRoleBindingByClusterRoleName, key)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	for _, obj := range objs {
+		crb := obj.(*rbacv1.ClusterRoleBinding)
+		c.enqueueClusterRoleBinding(crb, "reason", "ClusterRole", "ClusterRole.name", name)
+	}
+}
+
+// Start starts the controller, which stops when ctx.Done() is closed.
+func (c *controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logger := logging.WithReconciler(klog.FromContext(ctx), ControllerName)
+	ctx = klog.NewContext(ctx, logger)
+	logger.Info("Starting controller")
+	defer logger.Info("Shutting down controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func (c *controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	logger := logging.WithQueueKey(klog.FromContext(ctx), key)
+	ctx = klog.NewContext(ctx, logger)
+	logger.V(4).Info("processing key")
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", ControllerName, key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *controller) process(ctx context.Context, key string) error {
```

Comment:
change the signature of this method to `ctx context.Context, crb *rbacv1.ClusterRoleBinding`, rename to `reconcile` and move it to a separate file - it would unify with the rest of the controllers and allow for easier unit testing

Reply:
done